freezer/src/freezer_files.rs

```diff
@@ -0,0 +1,366 @@
+            .map_err(internal_error)?;
+        index.read_exact(&mut buffer).map_err(internal_error)?;
+        let end_index = IndexEntry::decode(&buffer)?;
+        if item == 1 {
+            return Ok((0, end_index.offset, end_index.file_id));
+        }
+
+        index
+            .seek(SeekFrom::Start((item - 1) * INDEX_ENTRY_SIZE))
+            .map_err(internal_error)?;
+        index.read_exact(&mut buffer).map_err(internal_error)?;
+        let start_index = IndexEntry::decode(&buffer)?;
+
+        if start_index.file_id != end_index.file_id {
+            return Ok((0, end_index.offset, end_index.file_id));
+        }
+
+        Ok((start_index.offset, end_index.offset, end_index.file_id))
+    }
+
+    fn preopen(&mut self) -> Result<(), Error> {
+        self.release_after(0);
+
+        for id in self.tail_id..self.head_id {
+            self.open_read_only(id)?;
+        }
+        self.files.insert(
+            self.head_id,
+            self.head.file.try_clone().map_err(internal_error)?,
+        );
+        Ok(())
+    }
+
+    fn write_index(&mut self, file_id: FileId, offset: u64) -> Result<(), Error> {
+        let index = IndexEntry { file_id, offset };
+        self.index
+            .write_all(&index.encode())
+            .map_err(internal_error)?;
+        Ok(())
+    }
+
+    fn release(&mut self, id: FileId) {
+        self.files.remove(&id);
+    }
+
+    fn release_after(&mut self, id: FileId) {
+        self.files.split_off(&id);
+    }
+
+    fn open_read_only(&mut self, id: FileId) -> Result<File, Error> {
+        let mut opt = fs::OpenOptions::new();
+        opt.read(true);
+        self.open_file(id, opt)
+    }
+
+    fn open_truncated(&mut self, id: FileId) -> Result<File, Error> {
+        let mut opt = fs::OpenOptions::new();
+        opt.create(true).read(true).write(true).truncate(true);
+        self.open_file(id, opt)
+    }
+
+    fn open_file(&mut self, id: FileId, opt: fs::OpenOptions) -> Result<File, Error> {
+        let name = helper::file_name(id);
+        let file = opt
+            .open(self.file_path.join(name))
+            .map_err(internal_error)?;
+        self.files
+            .insert(id, file.try_clone().map_err(internal_error)?);
+        Ok(file)
+    }
+}
+
+pub struct FreezerFilesBuilder {
+    file_path: PathBuf,
+    max_file_size: u64,
+}
+
+impl FreezerFilesBuilder {
+    pub fn new(file_path: PathBuf) -> Self {
+        FreezerFilesBuilder {
+            file_path,
+            max_file_size: MAX_FILE_SIZE,
+        }
+    }
+
+    pub(crate) fn max_file_size(mut self, size: u64) -> Self {
+        self.max_file_size = size;
+        self
+    }
+
+    pub fn build(self) -> Result<FreezerFiles, Error> {
+        fs::create_dir_all(&self.file_path).map_err(internal_error)?;
+        let mut index = self.open_index()?;
+
+        let index_meta = index.metadata().map_err(internal_error)?;
+        let mut index_size = index_meta.len();
```

Comment:
The size is not checked that is multiple of `INDEX_ENTRY_SIZE`.

Reply:
this already done when open_index